File: fixtures/src/OpStreamFuzz.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.stream.Stream;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field;
import org.apache.lucene.document.LongPoint;
import org.apache.lucene.document.NumericDocValuesField;
import org.apache.lucene.document.SortedDocValuesField;
import org.apache.lucene.document.StringField;
import org.apache.lucene.document.TextField;
import org.apache.lucene.index.DirectoryReader;
import org.apache.lucene.index.IndexWriter;
import org.apache.lucene.index.IndexWriterConfig;
import org.apache.lucene.index.LeafReader;
import org.apache.lucene.index.LeafReaderContext;
import org.apache.lucene.index.NumericDocValues;
import org.apache.lucene.index.PointValues;
import org.apache.lucene.index.SerialMergeScheduler;
import org.apache.lucene.index.SortedDocValues;
import org.apache.lucene.index.Term;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.PhraseQuery;
import org.apache.lucene.search.TermQuery;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.ScoreDoc;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.FSDirectory;
import org.apache.lucene.util.Bits;
import org.apache.lucene.util.BytesRef;

/**
 * The Java half of M4's differential operation-stream fuzzer (T4.5): the same seeded stream of
 * adds, updates, deletes (by id and by body word), doc-values updates, flushes and commits,
 * applied to real Lucene's {@code IndexWriter}, then dumped **semantically** -- live documents
 * with their doc values and points, the live documents matching every term and a set of
 * phrases -- so the dump is comparable with this port's regardless of how either engine laid out
 * its segments.
 *
 * <p>{@code crates/lucene-search/examples/op_stream_fuzz.rs} is the other half, with the same
 * generator; {@code scripts/op-stream-fuzz.sh} compares the two.
 *
 * <p>Usage: {@code OpStreamFuzz <out-dir> <first-seed> <end-seed> <ops>}: writes
 * {@code <out-dir>/<seed>.java.txt} for every seed in {@code [first, end)}.
 */
public class OpStreamFuzz {
  static final int WORDS = 10;
  static final int XS = 13;

  /** xorshift64 over a SplitMix64-spread seed; identical to the Rust side's {@code Rng}. */
  static final class Rng {
    long s;

    Rng(long seed) {
      long z = seed + 0x9e3779b97f4a7c15L;
      z = (z ^ (z >>> 30)) * 0xbf58476d1ce4e5b9L;
      z = (z ^ (z >>> 27)) * 0x94d049bb133111ebL;
      s = (z ^ (z >>> 31)) | 1;
    }

    long next() {
      s ^= s << 13;
      s ^= s >>> 7;
      s ^= s << 17;
      return s;
    }

    long below(long n) {
      return Long.remainderUnsigned(next(), n);
    }
  }

  static String body(long id, long v) {
    return "w" + ((id * 7 + v) % WORDS) + " w" + ((id + 3 * v) % WORDS) + " x" + (id % XS);
  }

  static String cat(long id, long v) {
    return (id + v) % 6 == 0 ? null : "c" + ((id + v) % 5);
  }

  static long pt(long id, long v) {
    return id * 3 - v;
  }

  static Document document(long id, long v) {
    Document d = new Document();
    d.add(new StringField("id", "i" + id, Field.Store.NO));
    d.add(new TextField("body", body(id, v), Field.Store.NO));
    d.add(new NumericDocValuesField("idn", id));
    d.add(new NumericDocValuesField("ver", v));
    d.add(new NumericDocValuesField("score", id * 10 + v));
    String c = cat(id, v);
    if (c != null) {
      d.add(new SortedDocValuesField("cat", new BytesRef(c)));
    }
    d.add(new LongPoint("pt", pt(id, v)));
    return d;
  }

  static void run(Path dirPath, long seed, int numOps) throws IOException {
    Rng rng = new Rng(seed);
    int maxBufferedDocs = 2 + (int) rng.below(20);
    // id -> version, for choosing an existing document exactly as the Rust side does.
    TreeMap<Long, Long> live = new TreeMap<>();
    long nextId = 0;
    IndexWriterConfig cfg = new IndexWriterConfig();
    cfg.setMaxBufferedDocs(maxBufferedDocs);
    cfg.setRAMBufferSizeMB(IndexWriterConfig.DISABLE_AUTO_FLUSH);
    cfg.setMergeScheduler(new SerialMergeScheduler());
    try (Directory dir = FSDirectory.open(dirPath);
        IndexWriter w = new IndexWriter(dir, cfg)) {
      for (int op = 0; op < numOps; op++) {
        long roll = rng.below(100);
        boolean needsPick = roll >= 48 && roll <= 77 && !(roll >= 67 && roll <= 70);
        Long pick = null;
        // One pick in five is any id ever issued -- possibly deleted; the rest
        // are a live id. Same draws, same order, as the Rust side.
        if (!needsPick) {
          pick = null;
        } else if (nextId > 0 && rng.below(5) == 0) {
          pick = rng.below(nextId);
        } else if (!live.isEmpty()) {
          long k = rng.below(live.size());
          pick = live.keySet().stream().skip(k).findFirst().get();
        }
        if (roll <= 47 || (roll <= 77 && !(roll >= 67 && roll <= 70) && pick == null)) {
          long id = nextId++;
          w.addDocument(document(id, 0));
          live.put(id, 0L);
        } else if (roll <= 59) {
          Long current = live.get(pick);
          long v = current == null ? 1 : current + 1;
          w.updateDocument(new Term("id", "i" + pick), document(pick, v));
          live.put(pick, v);
        } else if (roll <= 66) {
          w.deleteDocuments(new Term("id", "i" + pick));
          live.remove(pick);
        } else if (roll <= 70) {
          long word = rng.below(WORDS);
          w.deleteDocuments(new Term("body", "w" + word));
          live.entrySet()
              .removeIf(
                  e -> {
                    String[] ws = body(e.getKey(), e.getValue()).split(" ");
                    return ws[0].equals("w" + word) || ws[1].equals("w" + word);
                  });
        } else if (roll <= 77) {
          long value = rng.below(100_000);
          w.updateNumericDocValue(new Term("id", "i" + pick), "score", value);
        } else if (roll <= 83) {
          w.flush();
        } else {
          w.commit();
        }
      }
      w.commit();
    }
  }

  static void dump(Path dirPath, Path out) throws IOException {
    List<String> lines = new ArrayList<>();
    try (Directory dir = FSDirectory.open(dirPath);
        DirectoryReader reader = DirectoryReader.open(dir)) {
      // Global doc -> id, through the idn doc values.
      long[] idOf = new long[reader.maxDoc()];
      TreeMap<Long, String> docs = new TreeMap<>();
      for (LeafReaderContext ctx : reader.leaves()) {
        LeafReader leaf = ctx.reader();
        Bits liveDocs = leaf.getLiveDocs();
        NumericDocValues idn = leaf.getNumericDocValues("idn");
        NumericDocValues ver = leaf.getNumericDocValues("ver");
        NumericDocValues score = leaf.getNumericDocValues("score");
        SortedDocValues cat = leaf.getSortedDocValues("cat");
        Map<Integer, List<Long>> points = new TreeMap<>();
        PointValues pv = leaf.getPointValues("pt");
        if (pv != null) {
          pv.intersect(
              new PointValues.IntersectVisitor() {
                @Override
                public void visit(int docID) {
                  throw new IllegalStateException();
                }

                @Override
                public void visit(int docID, byte[] packed) {
                  points.computeIfAbsent(docID, d -> new ArrayList<>())
                      .add(LongPoint.decodeDimension(packed, 0));
                }

                @Override
                public PointValues.Relation compare(byte[] min, byte[] max) {
                  return PointValues.Relation.CELL_CROSSES_QUERY;
                }
              });
        }
        for (int doc = 0; doc < leaf.maxDoc(); doc++) {
          long id = idn.advanceExact(doc) ? idn.longValue() : -1;
          idOf[ctx.docBase + doc] = id;
          if (liveDocs != null && !liveDocs.get(doc)) {
            continue;
          }
          String v = ver.advanceExact(doc) ? Long.toString(ver.longValue()) : "-";
          String s = score.advanceExact(doc) ? Long.toString(score.longValue()) : "-";
          String c =
              cat != null && cat.advanceExact(doc)
                  ? cat.lookupOrd(cat.ordValue()).utf8ToString()
                  : "-";
          List<Long> p = points.getOrDefault(doc, List.of());
          if (docs.put(id, "doc " + id + " ver " + v + " score " + s + " cat " + c + " pt " + p)
              != null) {
            lines.add("DUPLICATE id " + id);
          }
        }
      }
      lines.addAll(docs.values());

      List<String> terms = new ArrayList<>();
      for (int i = 0; i < WORDS; i++) terms.add("w" + i);
      for (int i = 0; i < XS; i++) terms.add("x" + i);
      // Through a scored TermQuery, as the Rust side does.
      IndexSearcher searcher = new IndexSearcher(reader);
      searcher.setQueryCache(null);
      for (String t : terms) {
        lines.add("term " + t + " " + hits(searcher, new TermQuery(new Term("body", t)), idOf));
      }
      for (int a = 0; a < WORDS; a++) {
        for (int b = 0; b < WORDS; b += 3) {
          lines.add("phrase w" + a + " w" + b + " " + hits(searcher, new PhraseQuery("body", "w" + a, "w" + b), idOf));
        }
        lines.add("phrase w" + a + " x" + a + " " + hits(searcher, new PhraseQuery("body", "w" + a, "x" + a), idOf));
      }
      lines.add("range pt -20..150 " + hits(searcher, LongPoint.newRangeQuery("pt", -20, 150), idOf));
    }
    Files.writeString(out, String.join("\n", lines) + "\n", StandardCharsets.UTF_8);
    // Segment count and maxDoc: evidence that merges ran, beside the dump.
    try (Directory dir = FSDirectory.open(dirPath);
        DirectoryReader reader = DirectoryReader.open(dir)) {
      Path meta = out.resolveSibling(out.getFileName().toString().replace(".txt", ".meta"));
      Files.writeString(meta, reader.leaves().size() + " " + reader.maxDoc() + "\n");
    }
  }

  static TreeSet<Long> hits(IndexSearcher searcher, Query q, long[] idOf) throws IOException {
    TreeSet<Long> ids = new TreeSet<>();
    int n = Math.max(1, searcher.getIndexReader().maxDoc());
    for (ScoreDoc sd : searcher.search(q, n).scoreDocs) {
      ids.add(idOf[sd.doc]);
    }
    return ids;
  }

  static void delete(Path p) throws IOException {
    if (!Files.exists(p)) return;
    try (Stream<Path> walk = Files.walk(p)) {
      for (Path f : (Iterable<Path>) walk.sorted(Comparator.reverseOrder())::iterator) {
        Files.delete(f);
      }
    }
  }

  public static void main(String[] args) throws IOException {
    Path out = Path.of(args[0]);
    long first = Long.parseLong(args[1]);
    long end = Long.parseLong(args[2]);
    int numOps = Integer.parseInt(args[3]);
    Files.createDirectories(out);
    Path work = Files.createTempDirectory("op-stream-java");
    for (long seed = first; seed < end; seed++) {
      Path idx = work.resolve("s" + seed);
      run(idx, seed, numOps);
      dump(idx, out.resolve(seed + ".java.txt"));
      delete(idx);
    }
    delete(work);
    System.out.println("OpStreamFuzz: Lucene ran seeds " + first + ".." + end);
  }
}

File: crates/lucene-search/examples/op_stream_fuzz.rs
//! The Rust half of M4's differential operation-stream fuzzer (T4.5):
//! `fixtures/src/OpStreamFuzz.java`'s seeded stream -- adds, updates,
//! deletes by id and by body word, numeric doc-values updates, flushes and
//! commits -- applied to this port's `IndexWriter`, then dumped in the same
//! **semantic** form, so `scripts/op-stream-fuzz.sh` can compare the two
//! engines line by line. Segment counts, merge timing and file layouts
//! legitimately differ and do not appear in the dump.
//!
//! The dump: every live document (`doc <id> ver <v> score <s> cat <c> pt
//! [<p>]`, by id); for every body term, the ids of the live documents it
//! matches; for a set of two-word phrases and one point range, the ids
//! matched. This port's reader returns doc ids, not stored documents, so ids
//! come from the `idn` doc values, as on the Java side.
//!
//! Usage: `op_stream_fuzz <out-dir> <first-seed> <end-seed> <ops>` writes
//! `<out-dir>/<seed>.rust.txt` for every seed in `first..end`.
// Example code: see `docs/arithmetic-gate.md`'s "Test code" section.
#![allow(clippy::arithmetic_side_effects)]

use std::collections::{BTreeMap, BTreeSet};
use std::fmt::Write as _;
use std::path::Path;

use lucene_codecs::field_infos::{DocValuesType, FieldInfo, IndexOptions};
use lucene_codecs::points::{IntersectVisitor, Relation};
use lucene_codecs::stored_fields::{Document, FieldValue, StoredField};
use lucene_codecs::{doc_values, points, terms_dict};
use lucene_index::buffered_updates::Term;
use lucene_index::index_writer::IndexWriter;
use lucene_index::merge_policy::MergePolicyConfig;
use lucene_index::segment_info::LuceneVersion;
use lucene_search::directory_reader::DirectoryReader;
use lucene_search::query::{BooleanQuery, Clause, PhraseQuery, TermQuery};
use lucene_search::{search_boolean_query_multi_segment, search_term_query_multi_segment};
use lucene_store::FsDirectory;

const WORDS: u64 = 10;
const XS: u64 = 13;

/// Identical to `OpStreamFuzz.Rng`.
struct Rng(u64);

impl Rng {
    fn new(seed: u64) -> Self {
        let mut z = seed.wrapping_add(0x9e37_79b9_7f4a_7c15);
        z = (z ^ (z >> 30)).wrapping_mul(0xbf58_476d_1ce4_e5b9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94d0_49bb_1331_11eb);
        Rng((z ^ (z >> 31)) | 1)
    }

    fn next(&mut self) -> u64 {
        self.0 ^= self.0 << 13;
        self.0 ^= self.0 >> 7;
        self.0 ^= self.0 << 17;
        self.0
    }

    fn below(&mut self, n: u64) -> u64 {
        self.next() % n
    }
}

fn body(id: u64, v: u64) -> String {
    format!(
        "w{} w{} x{}",
        (id * 7 + v) % WORDS,
        (id + 3 * v) % WORDS,
        id % XS
    )
}

fn cat(id: u64, v: u64) -> Option<String> {
    (!(id + v).is_multiple_of(6)).then(|| format!("c{}", (id + v) % 5))
}

fn pt(id: u64, v: u64) -> i64 {
    id as i64 * 3 - v as i64
}

const F_ID: i32 = 0;
const F_BODY: i32 = 1;
const F_IDN: i32 = 2;
const F_VER: i32 = 3;
const F_SCORE: i32 = 4;
const F_CAT: i32 = 5;
const F_PT: i32 = 6;

fn schema() -> Vec<FieldInfo> {
    vec![
        FieldInfo {
            index_options: IndexOptions::Docs,
            omit_norms: true,
            ..FieldInfo::new("id", F_ID)
        },
        FieldInfo {
            index_options: IndexOptions::DocsAndFreqsAndPositions,
            ..FieldInfo::new("body", F_BODY)
        },
        FieldInfo {
            doc_values_type: DocValuesType::Numeric,
            ..FieldInfo::new("idn", F_IDN)
        },
        FieldInfo {
            doc_values_type: DocValuesType::Numeric,
            ..FieldInfo::new("ver", F_VER)
        },
        FieldInfo {
            doc_values_type: DocValuesType::Numeric,
            ..FieldInfo::new("score", F_SCORE)
        },
        FieldInfo {
            doc_values_type: DocValuesType::Sorted,
            ..FieldInfo::new("cat", F_CAT)
        },
        FieldInfo {
            point_dimension_count: 1,
            point_index_dimension_count: 1,
            point_num_bytes: 8,
            ..FieldInfo::new("pt", F_PT)
        },
    ]
}

fn document(id: u64, v: u64) -> Document {
    let field = |field_number, value| StoredField {
        field_number,
        value,
    };
    let mut fields = vec![
        field(F_ID, FieldValue::String(format!("i{id}"))),
        field(F_BODY, FieldValue::String(body(id, v))),
        field(F_IDN, FieldValue::Long(id as i64)),
        field(F_VER, FieldValue::Long(v as i64)),
        field(F_SCORE, FieldValue::Long((id * 10 + v) as i64)),
    ];
    if let Some(c) = cat(id, v) {
        fields.push(field(F_CAT, FieldValue::String(c)));
    }
    fields.push(field(F_PT, FieldValue::Long(pt(id, v))));
    Document { fields }
}

fn run(path: &Path, seed: u64, num_ops: usize) -> Result<(), String> {
    let mut rng = Rng::new(seed);
    let max_buffered_docs = 2 + rng.below(20) as i32;
    let mut live: BTreeMap<u64, u64> = BTreeMap::new();
    let mut next_id = 0u64;
    let dir = FsDirectory::open(path);
    let e = |e: lucene_index::index_writer::Error| e.to_string();
    let mut w = IndexWriter::open(
        &dir,
        schema(),
        "Lucene104",
        LuceneVersion {
            major: 10,
            minor: 5,
            bugfix: 0,
        },
    )
    .map_err(e)?;
    w.set_postings_field(Some("id")).map_err(e)?;
    w.add_postings_field("body").map_err(e)?;
    w.set_doc_values_field(Some("idn")).map_err(e)?;
    for f in ["ver", "score", "cat"] {
        w.add_doc_values_field(f).map_err(e)?;
    }
    w.add_points_field("pt").map_err(e)?;
    w.set_max_buffered_docs(max_buffered_docs).map_err(e)?;
    w.set_ram_buffer_size_mb(4096.0).map_err(e)?;
    // `TieredMergePolicy`'s shape, with a small floor so merges fire at these
    // sizes; Java's own schedule differs, which is the point of comparing
    // semantically.
    w.set_merge_policy(Some(MergePolicyConfig {
        floor_segment_size: 1 << 16,
        ..MergePolicyConfig::default()
    }));
    let id_term = |id: u64| Term::new("id", format!("i{id}").into_bytes());

    for _ in 0..num_ops {
        let roll = rng.below(100);
        let needs_pick = (48..=77).contains(&roll) && !(67..=70).contains(&roll);
        // One pick in five is any id ever issued -- possibly deleted, so an
        // update of it is a plain add and a delete or doc-values update of it
        // matches nothing. The rest are a live id.
        let pick = if !needs_pick {
            None
        } else if next_id > 0 && rng.below(5) == 0 {
            Some(rng.below(next_id))
        } else if !live.is_empty() {
            let k = rng.below(live.len() as u64) as usize;
            live.keys().nth(k).copied()
        } else {
            None
        };
        if roll <= 47 || (needs_pick && pick.is_none()) {
            let id = next_id;
            next_id += 1;
            w.add_document(document(id, 0)).map_err(e)?;
            live.insert(id, 0);
        } else if roll <= 59 {
            let id = pick.expect("picked");
            let v = live.get(&id).map_or(1, |v| v + 1);
            w.update_document(id_term(id), document(id, v)).map_err(e)?;
            live.insert(id, v);
        } else if roll <= 66 {
            let id = pick.expect("picked");
            w.delete_documents_by_term(&[id_term(id)]).map_err(e)?;
            live.remove(&id);
        } else if roll <= 70 {
            let word = rng.below(WORDS);
            let term = format!("w{word}");
            w.delete_documents_by_term(&[Term::new("body", term.clone().into_bytes())])
                .map_err(e)?;
            live.retain(|&id, &mut v| !body(id, v).split(' ').take(2).any(|t| t == term));
        } else if roll <= 77 {
            let value = rng.below(100_000);
            w.update_numeric_doc_value(id_term(pick.expect("picked")), "score", value as i64)
                .map_err(e)?;
        } else if roll <= 83 {
            w.flush().map_err(e)?;
        } else {
            w.commit().map_err(e)?;
        }
    }
    w.commit().map_err(e)?;
    Ok(())
}

struct AllPoints(Vec<(i32, i64)>);

impl IntersectVisitor for AllPoints {
    fn compare(&mut self, _min: &[u8], _max: &[u8]) -> Relation {
        Relation::CellCrossesQuery
    }
    fn visit(&mut self, _doc_id: i32) {
        unreachable!("every cell crosses the query");
    }
    fn visit_with_value(&mut self, doc_id: i32, packed: &[u8]) {
        let bits = u64::from_be_bytes(packed.try_into().expect("8-byte point"));
        self.0.push((doc_id, (bits ^ (1 << 63)) as i64));
    }
}

fn set(ids: &BTreeSet<i64>) -> String {
    let items: Vec<String> = ids.iter().map(i64::to_string).collect();
    format!("[{}]", items.join(", "))
}

fn dump(path: &Path) -> Result<String, String> {
    let dir = FsDirectory::open(path);
    let reader = DirectoryReader::open(&dir).map_err(|e| e.to_string())?;
    let mut out = String::new();
    let mut id_of: Vec<i64> = Vec::new();
    let mut docs: BTreeMap<i64, String> = BTreeMap::new();
    for segment in reader.segment_readers() {
        let fields = &segment.field_infos().fields;
        let number = |name: &str| fields.iter().find(|f| f.name == name).map(|f| f.number);
        let numeric = |name: &str, doc: i32| -> Result<Option<i64>, String> {
            let Some(n) = number(name) else {
                return Ok(None);
            };
            let Some((meta, data)) = segment.doc_values_for_field(n) else {
                return Ok(None);
            };
            let Some(entry) = meta.numeric_entry(n) else {
                return Ok(None);
            };
            doc_values::numeric_value(data, entry, doc).map_err(|e| e.to_string())
        };
        let cat_terms = match number("cat").and_then(|n| {
            let (meta, data) = segment.doc_values_for_field(n)?;
            Some((meta.sorted_entry(n)?, data))
        }) {
            Some((entry, data)) => Some((
                entry,
                data,
                terms_dict::decode_all_terms(data, &entry.terms).map_err(|e| e.to_string())?,
            )),
            None => None,
        };
        let mut points_by_doc = AllPoints(Vec::new());
        if let (Some((kdm, kdi, kdd)), Some(n)) = (segment.points_files(), number("pt")) {
            let reader = points::open(kdm, kdi, kdd, &segment.segment_id(), "")
                .map_err(|e| e.to_string())?;
            if reader.field(n).is_some() {
                reader
                    .intersect(n, &mut points_by_doc)
                    .map_err(|e| e.to_string())?;
            }
        }
        points_by_doc.0.sort_unstable();
        for doc in 0..segment.max_doc {
            let id = numeric("idn", doc)?.unwrap_or(-1);
            id_of.push(id);
            if segment.live_docs().is_some_and(|l| !l.get(doc as usize)) {
                continue;
            }
            let show = |v: Option<i64>| v.map_or("-".to_string(), |v| v.to_string());
            let c = match &cat_terms {
                Some((entry, data, terms)) => doc_values::sorted_ord(data, entry, doc)
                    .map_err(|e| e.to_string())?
                    .map_or("-".to_string(), |ord| {
                        String::from_utf8_lossy(&terms[ord as usize]).into_owned()
                    }),
                None => "-".to_string(),
            };
            let from = points_by_doc.0.partition_point(|&(d, _)| d < doc);
            let to = points_by_doc.0.partition_point(|&(d, _)| d <= doc);
            let p: Vec<String> = points_by_doc.0[from..to]
                .iter()
                .map(|&(_, v)| v.to_string())
                .collect();
            let line = format!(
                "doc {id} ver {} score {} cat {c} pt [{}]",
                show(numeric("ver", doc)?),
                show(numeric("score", doc)?),
                p.join(", ")
            );
            if docs.insert(id, line).is_some() {
                writeln!(out, "DUPLICATE id {id}").unwrap();
            }
        }
    }
    for line in docs.values() {
        writeln!(out, "{line}").unwrap();
    }
    std::fs::write(
        path.with_extension("meta"),
        format!("{} {}\n", reader.segment_readers().len(), id_of.len()),
    )
    .map_err(|e| e.to_string())?;

    let opened = reader.open_segments().map_err(|e| e.to_string())?;
    let segments = opened.as_open_segments();
    let top_n = id_of.len().max(1);
    let norms = reader.field_norms("body");
    let norms: Vec<Option<&_>> = norms.iter().map(Option::as_ref).collect();
    let fields = vec!["body".to_string()];
    let by_field = reader.field_norms_by_field(&fields);
    let bool_norms: Vec<Option<&_>> = by_field.iter().map(Some).collect();
    let ids = |hits: Vec<lucene_search::collector::ScoreDoc>| -> BTreeSet<i64> {
        hits.iter().map(|h| id_of[h.doc_id as usize]).collect()
    };
    let mut terms: Vec<String> = (0..WORDS).map(|i| format!("w{i}")).collect();
    terms.extend((0..XS).map(|i| format!("x{i}")));
    for t in &terms {
        let hits = search_term_query_multi_segment(
            &segments,
            &TermQuery {
                field: "body".to_string(),
                term: t.as_bytes().to_vec(),
            },
            &norms,
            top_n,
        )
        .map_err(|e| e.to_string())?;
        writeln!(out, "term {t} {}", set(&ids(hits))).unwrap();
    }
    let phrase = |a: &str, b: &str| -> Result<BTreeSet<i64>, String> {
        let query = BooleanQuery {
            must: vec![Clause::Phrase(PhraseQuery {
                field: "body".to_string(),
                terms: vec![a.as_bytes().to_vec(), b.as_bytes().to_vec()],
                slop: 0,
            })],
            ..Default::default()
        };
        search_boolean_query_multi_segment(&segments, &query, &bool_norms, top_n)
            .map(ids)
            .map_err(|e| e.to_string())
    };
    for a in 0..WORDS {
        for b in (0..WORDS).step_by(3) {
            let got = phrase(&format!("w{a}"), &format!("w{b}"))?;
            writeln!(out, "phrase w{a} w{b} {}", set(&got)).unwrap();
        }
        let got = phrase(&format!("w{a}"), &format!("x{a}"))?;
        writeln!(out, "phrase w{a} x{a} {}", set(&got)).unwrap();
    }
    // Through the port's own `PointValues.intersect` pruning walk
    // (`PointsReader::range_query`), not the per-document values above.
    let mut in_range: BTreeSet<i64> = BTreeSet::new();
    let mut doc_base = 0usize;
    for segment in reader.segment_readers() {
        let n = segment
            .field_infos()
            .fields
            .iter()
            .find(|f| f.name == "pt")
            .map(|f| f.number);
        if let (Some((kdm, kdi, kdd)), Some(n)) = (segment.points_files(), n) {
            let points = points::open(kdm, kdi, kdd, &segment.segment_id(), "")
                .map_err(|e| e.to_string())?;
            if points.field(n).is_some() {
                let lower = lucene_search::pack_i64(-20);
                let upper = lucene_search::pack_i64(150);
                for doc in points
                    .range_query(n, &lower, &upper)
                    .map_err(|e| e.to_string())?
                {
                    if segment.live_docs().is_none_or(|l| l.get(doc as usize)) {
                        in_range.insert(id_of[doc_base + doc as usize]);
                    }
                }
            }
        }
        doc_base += segment.max_doc as usize;
    }
    writeln!(out, "range pt -20..150 {}", set(&in_range)).unwrap();
    Ok(out)
}

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let out = Path::new(&args[0]);
    let first: u64 = args[1].parse().expect("first seed");
    let end: u64 = args[2].parse().expect("end seed");
    let num_ops: usize = args[3].parse().expect("ops");
    std::fs::create_dir_all(out).expect("out dir");
    let work = std::env::temp_dir().join(format!("op-stream-rust-{}", std::process::id()));
    for seed in first..end {
        let idx = work.join(format!("s{seed}"));
        let _ = std::fs::remove_dir_all(&idx);
        std::fs::create_dir_all(&idx).expect("index dir");
        let text = run(&idx, seed, num_ops)
            .and_then(|()| dump(&idx))
            .unwrap_or_else(|why| format!("ERROR {why}\n"));
        std::fs::write(out.join(format!("{seed}.rust.txt")), text).expect("write dump");
        // Segment count and `maxDoc` at the end: evidence that merges ran
        // (a `maxDoc` below the documents ever added means a merge reclaimed
        // deletes).
        if let Ok(meta) = std::fs::read(idx.with_extension("meta")) {
            std::fs::write(out.join(format!("{seed}.rust.meta")), meta).expect("write meta");
        }
        let _ = std::fs::remove_dir_all(&idx);
    }
    let _ = std::fs::remove_dir_all(&work);
    println!("op_stream_fuzz: this port ran seeds {first}..{end}");
}
